simulariumio/custom_converter.py

```diff
@@ -92,3 +119,276 @@ def _read(self, input_data: CustomData) -> Dict[str, Any]:
+                        + (
+                            len(V1_SPATIAL_BUFFER_STRUCT)
+                            - V1_SPATIAL_BUFFER_STRUCT.index("NSP")
+                        )
+                    )
+                    get_n_subpoints = False
+                    continue
+                # there should be a unique ID at this index, check for duplicate
+                uid = data[i]
+                if uid in uids:
+                    raise Exception(
+                        f"found duplicate ID {uid} in frame {t} at index {i}"
+                    )
+                uids.append(uid)
+                i += V1_SPATIAL_BUFFER_STRUCT.index("NSP") - 1
+                get_n_subpoints = True
+        return True
+
+    @staticmethod
+    def _format_timestep(number: float):
+        return float("%.4g" % number)
+
+    @staticmethod
+    def _determine_plot_reader(plot_type: str = "scatter") -> [PlotReader]:
+        """
+        Return the plot reader to match the requested plot type
+        """
+        if plot_type in SUPPORTED_PLOT_READERS:
+            return SUPPORTED_PLOT_READERS[plot_type]
+
+        raise UnsupportedPlotTypeError(plot_type)
+
+    def add_plot(
+        self,
+        data: [ScatterPlotData or HistogramPlotData] = {},
+        plot_type: str = "scatter",
+    ):
+        """
+        Add data to be rendered in a plot
+
+        Parameters
+        ----------
+        data: ScatterPlotData or HistogramPlotData
+            Loaded data for a plot.
+        plot_type: str
+            A string specifying which type of plot to render.
+            Current options:
+                'scatter' : a scatterplot with y-trace(s) dependent
+                    on an x-trace
+                'histogram' : a histogram with bars drawn at intervals
+                    over the range(s) of the data, with their height
+                    corresponding to the number of values in each interval
+            Default: 'scatter'
+        """
+        plot_reader_class = self._determine_plot_reader(plot_type)
+        self._data["plotData"]["data"].append(plot_reader_class().read(data))
+
+    @staticmethod
+    def _determine_filter(filter_type: str) -> [Filter]:
+        """
+        Return the filter to match the requested filter type
+        """
+        if filter_type in FILTERS:
+            return FILTERS[filter_type]
+
+        raise UnsupportedFilterTypeError(filter_type)
+
+    def apply_filters(self, params: List[FilterParams]):
+        """
+        Apply the given filter to the simularium data
+
+        Parameters
+        ----------
+        params: List[FilterParams]
+            a list of filter parameters, one for each filter to be applied
+        """
+        box_size = self._data["trajectoryInfo"]["size"]
+        plot_data = self._data["plotData"]
+        agent_data = AgentData.from_simularium_data(self._data)
+        for i in range(len(params)):
+            filter_class = self._determine_filter(params[i].name)
```

Comment:
Perhaps passing in the Filter itself rather than its string name will make this more extensible

Reply:
how should I pass both the filter and the parameters? Make the filter function itself one of the parameters? Or pass them both in?